### ISSUE TITLE
this.set() doesn't return the value returned by the computed property (2.14.1)

### ISSUE THREAD

**Issue by ffaubry** [emberjs/ember.js#15526]
This twiddle is demonstrating my issue:
https://ember-twiddle.com/103397293bfab3c11e3a8573ddc05857?openFiles=controllers.application.js%2C

I found the piece of code behind this issue: https://github.com/emberjs/ember.js/blob/dd884ab7175579dc42b8eda1b16a6d40fd833d6d/packages/ember-metal/lib/property_set.js#L53
```
  if (isDescriptor) { /* computed property */
    currentValue.set(obj, keyName, value);
  } else if (currentValue === undefined && 'object' === typeof obj && !(keyName in obj) &&
    typeof obj.setUnknownProperty === 'function') { /* unknown property */
    obj.setUnknownProperty(keyName, value);
  } else if (currentValue === value) { /* no change */
  } else {
    let meta = peekMeta(obj);
    propertyWillChange(obj, keyName, meta);

    if (MANDATORY_SETTER) {
      setWithMandatorySetter(meta, obj, keyName, value);
    } else {
      obj[keyName] = value;
    }

    propertyDidChange(obj, keyName, meta);
  }

  return value;
```

IMO, instead of returning the value passed, this function should collect the value from the various set functions called:

```
  let valueSet = value;
  if (isDescriptor) { /* computed property */
    valueSet = currentValue.set(obj, keyName, value);
  } else if (currentValue === undefined && 'object' === typeof obj && !(keyName in obj) &&
    typeof obj.setUnknownProperty === 'function') { /* unknown property */
    valueSet = obj.setUnknownProperty(keyName, value);
  } else if (currentValue === value) { /* no change */
  } else {
    let meta = peekMeta(obj);
    propertyWillChange(obj, keyName, meta);

    if (MANDATORY_SETTER) {
      valueSet = setWithMandatorySetter(meta, obj, keyName, value);
    } else {
      obj[keyName] = valueSet;
    }

    propertyDidChange(obj, keyName, meta);
  }

  return valueSet;
```

Is this a bug or working as designed?

**Comment by stefanpenner** [comment]
Question: did this behavior change recently?

**Comment by Serabe** [comment]
This is not an bug. `set` behaviour was changed pre 2.0 to match JavaScript setters semantics. The following code:

```
let a = {
  set hola(value) {
    return 5;
  }
};

console.log(a.hola = 1);
```

[logs `1` and not `5` as you can check in this jsbin](http://jsbin.com/citudosulo/edit?html,js,console).

Thank you!

**Comment by ffaubry** [comment]
@stefanpenner I wrote recently a piece of code that was:
`return this.set('foo', value);`
expecting that it would return the value set and not the value passed. Because I know too much about computed property in which my setters return the value set...
Based on your answers, I verified the doc and it is matching what you said:
https://emberjs.com/api/ember/2.14/namespaces/Ember/methods/set?anchor=set

@Serabe My bad. Thanks for taking the time to clarify.
